### PR TITLE
[Issue-2101] Minimize finalized state regeneration

### LINKED ISSUE #2101: Don't needlessly generate finalized states
### Description
As of PR #2065, we're potentially regenerating a lot of finalized states to send to storage.  We should instead send only the immediately available finalized states to storage, and have the storage layer regenerate any states that it actually needs.  See: https://github.com/PegaSysEng/teku/blob/1087e3efada98af844c0137200b6af99643834dc/storage/src/main/java/tech/pegasys/teku/storage/store/StoreTransactionUpdates.java#L152-L163

### PR DESCRIPTION
**Pull request by mbaxter**
## PR Description
Move handling of newly finalized states to the `Database`.  This way, we won't bother regenerating finalized states if the storage mode is set to `PRUNE`. 

## Fixed Issue(s)
Fixes #2101 

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: ethereum/core/src/test/java/tech/pegasys/teku/core/StateGeneratorTest.java [hunk 1]
```diff
@@ -215,7 +230,19 @@ private void testRegenerateAllStates(
     assertThat(resultMap.size()).isEqualTo(results.size());
     // Check that our expectations are met
     assertThat(resultMap.size()).isEqualTo(expectedResult.size());
-    assertThat(resultMap).isEqualToComparingFieldByField(expectedResult);
+    assertThat(resultMap).containsExactlyInAnyOrderEntriesOf(expectedResult);
+    // Check states were / were not regenerated as expected
+    if (supplyAllKnownStates) {
+      // No states should be regenerated - they should all match the known state
+      for (Bytes32 root : expectedResult.keySet()) {
+        assertThat(resultMap.get(root) == expectedResult.get(root)).isTrue();
```
Comment (ajsutton) [inline ethereum/core/src/test/java/tech/pegasys/teku/core/StateGeneratorTest.java]:
```suggestion
        assertThat(resultMap.get(root)).isSameAs(expectedResult.get(root));
```

File: ethereum/core/src/test/java/tech/pegasys/teku/core/StateGeneratorTest.java [hunk 2]
```diff
@@ -215,7 +230,19 @@ private void testRegenerateAllStates(
...
+      }
+    } else if (cacheSize == 0) {
+      // All states should be regenerated and should not match the known states
+      for (Bytes32 root : expectedResult.keySet()) {
+        assertThat(resultMap.get(root) == expectedResult.get(root)).isFalse();
```
Comment (ajsutton) [inline ethereum/core/src/test/java/tech/pegasys/teku/core/StateGeneratorTest.java]:
```suggestion
        assertThat(resultMap.get(root)).isNotSameAs(expectedResult.get(root));
```

File: storage/src/main/java/tech/pegasys/teku/storage/store/StoreTransactionUpdates.java [hunk 3]
```diff
@@ -149,17 +154,18 @@ public static StoreTransactionUpdates calculate(final Store baseStore, final Tra
...
-              final BeaconState state = tx.getBlockState(block.getRoot());
-              return new SignedBlockAndState(block, state);
-            })
-        .collect(Collectors.toMap(SignedBlockAndState::getRoot, Function.identity()));
+    return blocks;
+  }
+
+  private static Map<Bytes32, BeaconState> collectFinalizedStates(
+      final Store.Transaction tx, final Set<SignedBeaconBlock> finalizedBlocks) {
+    final HashMap<Bytes32, BeaconState> states = new HashMap<>();
```
Comment (ajsutton) [inline storage/src/main/java/tech/pegasys/teku/storage/store/StoreTransactionUpdates.java]:
```suggestion
    final Map<Bytes32, BeaconState> states = new HashMap<>();
```

**Review by ajsutton — APPROVED** [review]
LGTM.

File: ethereum/core/src/test/java/tech/pegasys/teku/core/StateGeneratorTest.java (hunk 1 above)
Comment (mbaxter) [reply ethereum/core/src/test/java/tech/pegasys/teku/core/StateGeneratorTest.java]: thanks - I knew there was an API for this but couldn't find it!

**Pull request merged**